plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreModule.java

```diff
@@ -43,32 +46,53 @@ public void configure(Binder binder)
         newOptionalBinder(binder, HiveMetastoreDecorator.class);
         newExporter(binder).export(HiveMetastore.class)
                 .as(generator -> generator.generatedNameOf(CachingHiveMetastore.class));
+        newSetBinder(binder, Procedure.class).addBinding().toProvider(FlushHiveMetastoreCacheProcedure.class).in(Scopes.SINGLETON);
     }
 
     @Provides
     @Singleton
-    public HiveMetastore createCachingHiveMetastore(
+    public Optional<CachingHiveMetastore> createCachingHiveMetastore(
             NodeManager nodeManager,
             @ForCachingHiveMetastore HiveMetastore delegate,
             CachingHiveMetastoreConfig config,
             CatalogName catalogName,
             Optional<HiveMetastoreDecorator> hiveMetastoreDecorator)
     {
-        HiveMetastore decoratedDelegate = hiveMetastoreDecorator
-                .map(decorator -> decorator.decorate(delegate))
```

Comment:
Previously, the decorator was applied always, now it's applied on coordinator only, and only if caching is enabled.
This looks wrong... unless it's applied in the "otherwise" case in the other method. 
cc @kokosing

Reply:
It's decorated in both cases, but as you described in https://github.com/trinodb/trino/pull/10251#discussion_r770341637 it its not so clear.
Let me try to think how to refactor.

Reply:
Refactored to single Guice factory method